mockingbird-processor/src/main/kotlin/com/careem/mockingbird/processor/SpyGenerator.kt

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright Careem, an Uber Technologies Inc. company
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.careem.mockingbird.processor
+
+import com.google.devtools.ksp.processing.KSPLogger
+import com.google.devtools.ksp.processing.Resolver
+import com.google.devtools.ksp.symbol.KSClassDeclaration
+import com.google.devtools.ksp.symbol.KSFunctionDeclaration
+import com.google.devtools.ksp.symbol.KSPropertyDeclaration
+import com.google.devtools.ksp.symbol.KSType
+import com.google.devtools.ksp.symbol.KSTypeArgument
+import com.google.devtools.ksp.symbol.KSTypeParameter
+import com.squareup.kotlinpoet.FunSpec
+import com.squareup.kotlinpoet.MemberName
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.ksp.toClassName
+import com.squareup.kotlinpoet.ksp.toTypeName
+
+class SpyGenerator(
+    resolver: Resolver,
+    logger: KSPLogger,
+    functionsMiner: FunctionsMiner
+) : Generator(resolver, logger, functionsMiner) {
+
+    override fun resolveSupertype(): String = "Spy"
+
+    override fun decorateConstructor(classToMock: KSType, classBuilder: TypeSpec.Builder) {
+        val className = classToMock.toTypeName()
+        val propertyName = (classToMock.declaration as KSClassDeclaration).toClassName().simpleName.replaceFirstChar(Char::lowercase)
+        classBuilder
+            .primaryConstructor(
+                FunSpec.constructorBuilder()
+                    .addParameter(propertyName, className)
+                    .build()
+            )
+            .addProperty(
+                PropertySpec.builder(propertyName, className)
+                    .initializer(propertyName)
+                    .build()
+            )
+    }
+
+    override fun decorateFunctionBody(
```

Comment:
function body is different between mocks and spies, this is the spy impl for the code gen of that block